### PR TITLE
feat[UI - Wallet Stability] Create generic/reusable assets listview component

### LINKED ISSUE #14716: [UI - Wallet Stability] Create generic/reusable assets listview component
Create a new generic and reusable listview component and delegates used for assets (to replace `TokenListView`):

  - Called “list” but it’s not a list but a composition of some components + 2 different lists. It could be decomposed into lists itself and another top level component composing lists together using tabs
  - In related *storybook page* there are done a lot of data transformations, some dependency of `TransactionStore` / `WalletAssetsStore`  only to transform assets model data. This should be transparent for the storybook page itself. The storybook page for given component shouldn’t bother about earlier transformation we do in the app. It’s sufficient to provide just what’s expected by the component, in a form as easy as ListModel with static ListElement items. Then in storybook we clearly see what the structure of the model expected by the component. Having it done this way, we don’t need to copy logic from real store to stub store. Stub stores should be just empty QtObjects.
  - `TransactionStore` / `WalletAssetsStore` contains a lot of UI-side logic related to data transformation. It should be excluded from stores and kept in a separate components (Adaptors - responsible for that data transformation). Stores should be only a thin layer exposing backend’s models/properties/methods.
  - The component takes function for formatting balances, (`formatCurrentCurrencyAmount`, `formatCurrencyAmountFromBigInt`). In storybook page it could be very simplified implementation, like e.g. `Math.round(balance) + " USD`. No need to involve stores. Ideally, models taken by the component should contain ready-to-display data, with no need to use external methods for formatting
  - Instead of taking formatting info from `CurrenciesStore` via a method, we could have this info exposed as a model. Then on UI side left join may be performed to combine assets and corresponding formatting info. Other solution (probably more efficient) is to expose formatting details directly via assets model.
  - `TokenListView` exposes `assetSearchString` , the assumption is that model’s data are filtered externally according to that string. It makes that component not fully self-contained. When instantiated in storybook search doesn’t work. This type of search could be implementation detail of that component - encapsulated and hidden.
  - `TokenListView` takes networks model in order to do left joins internally for submodel of assets and collectibles. Currently, with the `SubmodelProxyModel` it’s no longer needed. That join may be done externally on submodels. It’s no longer needed to have network model as a dependency.
  - All model transformations, like adding role `"isCommunityAsset"` should be done externally. UI component should just depend on a model with such role.

### PR DESCRIPTION
**Pull request by caybro**
### What does the PR do

TLDR: later this should form a basic building block for a new TokenSelector picker component, potentially replacing the current HoldingSelector* and TokenListView components (support for collectibles TBD as part of https://github.com/status-im/status-desktop/issues/15121)

- create reusable `TokenSelectorAssetDelegate` and `TokenSelectorView`
- add corresponding SB page, showcasing the flow/integration and the separation of concerns between the view, adaptor and delegate layers
- add QML testcase for TokenSelectorView
- don't display crypto symbol for token balances per chain tags
- update the stores and SB pages
- add some missing formatter functions to LocaleUtils and CurrenciesStore

Fixes #14716

### Affected areas

TokenSelectorView, TokenSelectorAssetDelegate

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

image

### CODE REVIEW

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+      Computed values:
+      - currentCurrencyBalance: double (e.g. `1000.42` in user's fiat currency)
+      - currentBalance: double (e.g. `1.42`)
+    */
+
+    // input API
+    required property var assetsModel
+    required property var flatNetworksModel
+    required property CurrenciesStore currencyStore
```
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:35]: I strongly recommend delaying passing the whole stores as long it's reasonable. This store has tens of methods and props so reading API of `TokenSelectorViewAdaptor` now doesn't really says what is needed by this component. It highly violates `ISP` of `SOLID`. Pushing dependency on stores up in the hierarchy will help us with better "layering" of the app and simpler testing.

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    // optional filter properties; empty/default values means no filtering
+    property var enabledChainIds: []
+    property string accountAddress
+    property bool showCommunityAssets
+    property string searchString
```
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:41]: Search functionality shouldn't be part of this adaptor but the component providing search bar. Search in this case is local thing, which can be done internally and don't have to be exposed.

File: storybook/pages/TokenSelectorViewPage.qml [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+import SortFilterProxyModel 0.2
+
+import AppLayouts.Wallet.views 1.0
+import AppLayouts.Wallet.stores 1.0
+import AppLayouts.Wallet.adaptors 1.0
+
+import shared.stores 1.0
+import utils 1.0
+
+SplitView {
```
Comment (micieslak) [inline storybook/pages/TokenSelectorViewPage.qml:24]: The storybook page, similarly as unit test, should be focused on a single "unit" of code, in this case `TokensSelectorView`. Therefore is shouldn't test or depend on correctness of other components like adaptor, which should be fully independent. I think the good idea is to have separate pages for views and adaptors. We already have some utility storybook components to easily display generic lists, they can be used to visualize input and output of adaptor withing a dedicated pages. Such page should contain controls for switches exposed from the adaptor.

Later, while progressing with refactors we will have sb pages combining both views and stores under more high-level components.

File: ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml [hunk 4]
```diff
@@ -22,6 +22,8 @@ QObject {
     property bool swapProposalReady: false
     property bool swapProposalLoading: false
 
+    property bool showCommunityTokens
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:25]: where does this get assigned? we should probably get that store here in that case.

File: ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml [hunk 5]
```diff
@@ -203,7 +201,7 @@ QObject {
             }
             readonly property LeftJoinModel joinModel: LeftJoinModel {
                 leftModel: submodel
-                rightModel: root.filteredFlatNetworksModel
+                rightModel: root.swapStore.flatNetworks
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:204]: why not use the root.filteredFlatNetworksModel instead which filters testnet/mainnet networks already?

File: ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml [hunk 6]
```diff
@@ -227,11 +225,12 @@ QObject {
     }
 
     /* Internal function to calculate total balance */
-    function __getTotalBalance(balances, decimals) {
+    function __getTotalBalance(balances, decimals, chainIds = [root.swapFormData.selectedNetworkChainId]) {
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:228]: ideally there must be another way to calculate total balance now? using the SubModelProxyModel ?

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                // let the root handle the click
+                TapHandler {
+                    onTapped: root.clicked()
+                }
+            }
+        }
+    }
+
+    onClicked: {
+        console.warn("!!! CLICKED INDEX:", index)
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml:129]: leftover remove?

File: ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml (hunk 5 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:204]: It doesn't work correctly that way, the `LeftJoinModel` is not filtering so you'll get ghost/incorrect entries that should not be there; I'll fix it as part of the followup PR here: https://github.com/status-im/status-desktop/issues/14783

File: ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml (hunk 6 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:228]: Yeah, same comment as above, I'll get rid of this completely

**Review by micieslak — COMMENTED** [review]
Looks great in general! I believe our new architecture starts emerging from chaos :D Once we are on the beginning of that way I'm proposing some alignments which are suitable here and easier to apply now, especially that now we are creating some kind of examples to be repeated in the future not only by ui-team.

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    id: root
+
+    required property string tokensKey
+    required property string name
+    required property string symbol
+    required property double currentCurrencyBalance
+    // expected structure: balancesModel -> model.balances submodel [chainId: int, balance: BigIntString] + flatNetworks [account:string, iconUrl: string]
+    required property var balancesModel
+
+    property var formatCurrencyAmount: (balance) => "IMPLEMENT ME"
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml:21]: do you want to print this or?

File: ui/StatusQ/src/objectproxymodel.cpp [hunk 9]
```diff
@@ -300,30 +298,32 @@ void ObjectProxyModel::updateIndexes(int from, int to)
...
-    for (auto& role : expectedRoles) {
+    for (auto &role : expectedRoles) {
         auto expectedKeys = roleNames.keys(role.toUtf8());
         auto expectedKeysCount = expectedKeys.size();
 
         if (expectedKeysCount == 1)
             expected.insert(expectedKeys.first(), role.toUtf8());
         else if (expectedKeysCount == 0) {
-            qWarning() << "Expected role not found!";
+            qWarning() << Q_FUNC_INFO;
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/StatusQ/src/objectproxymodel.cpp:321]: Yeah, and that's why I did it this way in order not to interfere too much with the test; I noticed you did something similar inf your PR. My point here is that while I was hunting an issue in my code, I really wanted to know which role is triggering the warning :) (the line below this one)

**Review by Khushboo-dev-cpp — COMMENTED** [review]
hey awesome work overall, just a few doubts - 
1. Already selected token view ? https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3504-426117&m=dev 
2. What about inline headers for the items like mentioned here ? I don’t know what the decision from Ben was do we want popular assets etc? Then do we add support for that already here ? https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3406-236702&m=dev
3. In case non balances are present on given network do we should 0 balance with each chainId or not show anything as done currently?

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+                }
+            }
+        }
+    }
+
+    onClicked: {
+        console.warn("!!! CLICKED INDEX:", index)
+        ListView.view.currentIndex = index
+        root.assetSelected(root.tokensKey)
+    }
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml:132]: Yeah, I don't really need the `currentIndex` as it turned out to be useless for the ComboBox integration; will remove

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+                    height: ListView.view.height
+                    spacing: 4
+                    StatusRoundedImage {
+                        Layout.preferredWidth: 16
+                        Layout.preferredHeight: 16
+                        image.source: Style.svg("tiny/%1".arg(model.iconUrl))
+                    }
+                    StatusBaseText {
+                        font.pixelSize: Theme.tertiaryTextFontSize
+                        text: LocaleUtils.numberToLocaleString(model.balanceAsDouble, 2)
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml:117]: You're right... this should be formatted differently (esp. for very low or large amounts); I'm not keen on externalizing it too much though :)

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 2 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:41]: I see no reason to exclude it; after all it's yet another filter parameter, just like the ones above (`accountAddress` or `enabledChainIds`)

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml (hunk 8 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml:21]: No, that's a hint for the developer/user that this method is really meant to be overridden (because we can't make it `required`)

**Comment by caybro** [comment]
> hey awesome work overall, just a few doubts -
> 
> 1. Already selected token view ? [https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3504-426117&m=dev    ](https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3504-426117&m=dev%E2%80%A8)

Yeah, will take care of that in the followup PR

> 2. What about inline headers for the items like mentioned here ? I don’t know what the decision from Ben was do we want popular assets etc? Then do we add support for that already here ? https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3406-236702&m=dev

Not sure about the actual idea behind this; will double check with him; but in any case, I'd say a ListView should stay a list view; we can add the sections later on in the actual `TokenSelector`

> 3. In case non balances are present on given network do we should 0 balance with each chainId or not show anything as done currently?

I'd say do not show anything at all, correct @benjthayer

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 1 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:35]: Yeah, not a big fan of passing the whole store around; I'll narrow it down to the only 2 methods and 1 property that we actually need from there

File: ui/app/AppLayouts/Wallet/views/TokenSelectorView.qml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+import AppLayouts.Wallet.adaptors 1.0
+
+import SortFilterProxyModel 0.2
+
+StatusListView {
+    id: root
+
+    // input API
+    required property TokenSelectorViewAdaptor adaptor
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorView.qml:16]: I only followed the same approach with the `SwapModal`; there the whole popup (view) also depends on the given adaptor and I kinda liked the approach. The adaptor should not change, and imho they should not serve multiple views. 

The dependency on the stores... yeah, I don't like this either and I'll drop it.

On a second thought... the `ListView` itself doesn't really need it, it's more the delegate that expects a certain set of roles to be available; I'll give it a second thought :)

File: ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml (hunk 4 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:25]: it's an optional filter; so it might get assigned from the respective view/component as needed

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+import utils 1.0
+
+ItemDelegate {
+    id: root
+
+    required property string tokensKey
+    required property string name
+    required property string symbol
+    required property double currentCurrencyBalance
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml]: Fixed

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml (hunk 11 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml]: Fixed

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml (hunk 8 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml]: Fixed

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml (hunk 10 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml]: Removed

File: ui/app/AppLayouts/Wallet/views/TokenSelectorView.qml (hunk 12 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorView.qml]: Removed the dependency

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 1 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml]: Removed the store from the adaptor

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+                    roleName: "symbol"
+                    pattern: root.searchString
+                    caseSensitivity: Qt.CaseInsensitive
+                }
+            },
+            ValueFilter {
+                roleName: "isCommunityAsset"
+                value: false
+                enabled: !root.showCommunityAssets
+            }
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml]: Right, removed

**Comment by caybro** [comment]
@Khushboo-dev-cpp @micieslak I believe I addressed all your comments, pls have a second look

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml (hunk 8 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml]: And I removed it now :)

File: ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml (hunk 7 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/views/TokenSelectorAssetDelegate.qml]: Removed

File: storybook/qmlTests/tests/tst_TokenSelectorView.qml [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+            currentCurrency: d.currencyStore.currentCurrency
+        }
+    }
+
+    Component {
+        id: componentUnderTest
+        TokenSelectorView {
+            anchors.fill: parent
+
+            model: d.adaptor.outputAssetsModel
```
Comment (micieslak) [inline storybook/qmlTests/tests/tst_TokenSelectorView.qml:43]: The goal of unit tests is to test components in isolation. It can be easily achieved once adaptor is decoupled from the view. Instead of using adaptor here we can use just plain `ListModel` here:
```
model: ListModel { /* test data here */ }
```

Similar rules regarding isolation should be followed in storybook.

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 2 above)
Comment (micieslak) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:40]: From architectural point of view there is a reason. It's not a problem for `TokenSelectorView` which is just a list view, but the component combining that one and search bar. This component can (and should) keep search as own details, there is no need to externalize that functionality from that component. Then the component is self-contained and own the logic it offers via search text input.

**Review by micieslak — COMMENTED** [review]
There are some points related to the app architecture that requires further adjustments. View and store are already separated in components code, but there is still better separation needed in unit test and storybook pages. What maybe now looks like minor thing will be important when we go up in components hierarchy to more complex one.

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+            readonly property int decimals: model.decimals
+            readonly property double currentBalance: aggregator.value
+            readonly property double currencyBalance: {
+                if (!!model.marketDetails) {
+                    return currentBalance * model.marketDetails.currencyPrice.amount
+                }
+                return 0
+            }
+            readonly property string currencyBalanceAsString:
+                LocaleUtils.currencyAmountToLocaleString({amount: currencyBalance, symbol: root.currentCurrency, displayDecimals: 2})
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:93]: Yeah, but here the design is very explicit about the decimals :thinking: 

I'll give it a second thought :)

File: storybook/qmlTests/tests/tst_TokenSelectorView.qml [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+            d.adaptor.accountAddress = "0x7F47C2e98a4BBf5487E6fb082eC2D9Ab0E6d8881"
+            waitForItemPolished(controlUnderTest)
+
+            // grab the "STT" delegate
+            const delegate = findChild(controlUnderTest, "tokenSelectorAssetDelegate_STT")
+            verify(!!delegate)
+
+            // should have 45.90 balance
+            tryCompare(delegate, "currencyBalanceAsString", "45.90 USD")
+        }
```
[the remark this answers is not in the document]
Comment (caybro) [reply storybook/qmlTests/tests/tst_TokenSelectorView.qml:147]: OK, I can split them

File: storybook/qmlTests/tests/tst_TokenSelectorView.qml (hunk 15 above)
Comment (caybro) [reply storybook/qmlTests/tests/tst_TokenSelectorView.qml:43]: This is just a convenience, the view is now properly separated from the adaptor; I'm just using its result as the view's model

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 16 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:93]: Also, I really want to keep the implementation details on _how_ to format a number inside the adaptor for many reasons:
- we decided we don't like things like `property var formatFoo: (balance) => "BAR"` in the API; this gets rid of that
- ppl simply tend to forget to override this formatter function unless they are forced to
- last but not least, we have all the information here inside the adaptor to provide that extra string role, so why externalize it

[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml]: Pushed a fix for the decimal places

File: ui/StatusQ/tests/tst_ObjectProxyModel.cpp [hunk 18]
```diff
@@ -263,9 +263,10 @@ private slots:
         model.setSourceModel(sourceModel);
         model.setDelegate(delegate.get());
 
-        QTest::ignoreMessage(QtWarningMsg, "Expected role not found!");
+        const auto expRoleName = QStringLiteral("undefined");
+        QTest::ignoreMessage(QtWarningMsg, QStringLiteral("Expected role \"%1\" not found!").arg(expRoleName).toLatin1());
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/StatusQ/tests/tst_ObjectProxyModel.cpp:269]: Fixed

File: storybook/qmlTests/tests/tst_TokenSelectorView.qml (hunk 17 above)
[the remark this answers is not in the document]
Comment (caybro) [reply storybook/qmlTests/tests/tst_TokenSelectorView.qml]: Tests split

File: storybook/qmlTests/tests/tst_TokenSelectorViewAdaptor.qml [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    Component {
+        id: componentUnderTest
+        TokenSelectorViewAdaptor {
+            assetsModel: d.assetsStore.groupedAccountAssetsModel
+            flatNetworksModel: d.flatNetworks
+            currentCurrency: d.currencyStore.currentCurrency
+        }
+    }
```
Comment (micieslak) [inline storybook/qmlTests/tests/tst_TokenSelectorViewAdaptor.qml:40]: Similarly here, what is needed are simple models for `assetsModel` and `flatNetworksModel`. No need for stores in unit test of adaptor.

File: storybook/qmlTests/tests/tst_TokenSelectorView.qml [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    Component {
+        id: componentUnderTest
+        TokenSelectorView {
+            anchors.fill: parent
+
+            model: d.adaptor.outputAssetsModel
+        }
+    }
```
Comment (micieslak) [inline storybook/qmlTests/tests/tst_TokenSelectorView.qml:45]: What is needed here is just simple list model done according to the specification in the component's API: https://github.com/status-im/status-desktop/pull/15157/files#diff-3de6ffcec9aefd08e1e130e2044ed30fb4d7502ce958ae568faa168ff07d694cR9
```
 Component {
        id: componentUnderTest
        TokenSelectorView {
            anchors.fill: parent

            model: ListModel {
                   ListElement { <some content here> } 
            }
        }
    }
```

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+            readonly property double currentBalance: aggregator.value
+            readonly property double currencyBalance: {
+                if (!!model.marketDetails) {
+                    return currentBalance * model.marketDetails.currencyPrice.amount
+                }
+                return 0
+            }
+            readonly property string currencyBalanceAsString:
+                currencyBalance ? LocaleUtils.currencyAmountToLocaleString({amount: currencyBalance, symbol: root.currentCurrency,
+                                                                               displayDecimals: model.marketDetails.currencyPrice.displayDecimals})
```
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:94]: In one place `model.marketDetails` is checked (currencyBalance) and there is not. So it should be checked in both places or check skipped consistently.

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+            // properties exposed as roles to the top-level model
+            readonly property int decimals: model.decimals
+            readonly property double currentBalance: aggregator.value
+            readonly property double currencyBalance: {
+                if (!!model.marketDetails) {
+                    return currentBalance * model.marketDetails.currencyPrice.amount
+                }
+                return 0
+            }
+            readonly property string currencyBalanceAsString:
```
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:92]: It's a bit better because it provides correct behavior, but the displayDecimals and formatting rule in general is constant for all entries in the model. So there is no need to expect additional field in the model per row (model.marketDetails.currencyPrice.displayDecimals) and expose additional role. Something changing across entries should be passed via the model, fixed stuff can be passed via a single property, directly to the view, no need to go via adaptor. For that reason formatting function passed to the view is good option here.

**Review by micieslak — COMMENTED** [review]
Thanks for separating unit tests. They need to be also cleaned from unnecessary stuff. The last important step here is to separate sb pages. If needed, just cherry-pick needed commits from https://github.com/status-im/status-desktop/pull/15125 if they are handy also here.

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 21 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:94]: That's a recent design requirement (from yesterday's meeting) to show "nothing" instead of "0.00 USD"

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 22 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:92]: It is not, check my answer above

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 21 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:94]: image

Ie, "AAVE" would normally show "0.00 USD" but it should not since it doesn't have any balance for the selected account/network

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 22 above)
Comment (micieslak) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml:92]: I mean the same value for all items in the model. From perspective of the model it's just single param.

File: ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml (hunk 21 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/adaptors/TokenSelectorViewAdaptor.qml]: Right, fixed!

**Comment by caybro** [comment]
@micieslak pushed a separate SB page for the delegate only

**Review by micieslak — APPROVED** [review]
Thanks!
